mlir-clang/Lib/clang-mlir.cc

```diff
@@ -1530,8 +1529,125 @@ const clang::FunctionDecl *MLIRScanner::EmitCallee(const Expr *E) {
   return nullptr;
 }
 
-ValueWithOffsets MLIRScanner::VisitCallExpr(clang::CallExpr *expr) {
-  if (auto ic = dyn_cast<ImplicitCastExpr>(expr->getCallee()))
+std::pair<ValueWithOffsets, bool>
+MLIRScanner::EmitGPUCallExpr(clang::CallExpr *expr) {
+  if (auto ic = dyn_cast<ImplicitCastExpr>(expr->getCallee())) {
+    if (auto sr = dyn_cast<DeclRefExpr>(ic->getSubExpr())) {
+      if (sr->getDecl()->getIdentifier() &&
+          sr->getDecl()->getName() == "__syncthreads") {
+        builder.create<mlir::NVVM::Barrier0Op>(loc);
+        return make_pair(ValueWithOffsets(), true);
+      }
+      if (sr->getDecl()->getIdentifier() &&
+          sr->getDecl()->getName() == "__log2f") {
+        std::vector<mlir::Value> args;
+        for (auto a : expr->arguments()) {
+          args.push_back(Visit(a).getValue(builder));
+        }
+        return make_pair(
+            ValueWithOffsets(builder.create<mlir::math::Log2Op>(loc, args[0]),
+                             /*isReference*/ false),
```

Comment:
I'm down for more organization, but this isn't a log2op for example.

Perhaps extract out the code into handle builtin ops, rather than handle gpu ops?